qemu/tests/qemu_guest_agent.py

```diff
@@ -935,24 +1037,30 @@ def setup(self, test, params, env):
             vm = self.env.get_vm(params["main_vm"])
             vm.verify_alive()
             self.vm = vm
-            session = self._get_session(params, self.vm)
+        session = self._get_session(params, self.vm)
 
-            if self._check_ga_pkg(session, params.get("gagent_pkg_check_cmd")):
-                logging.info("qemu-ga is already installed.")
-            else:
-                logging.info("qemu-ga is not installed.")
-                self.setup_gagent_in_host(session, params, self.vm)
-                self.gagent_install(session, self.vm, *[params.get("gagent_install_cmd")])
+        if self._check_ga_pkg(session, params.get("gagent_pkg_check_cmd")):
+            logging.info("qemu-ga is already installed.")
+        else:
+            logging.info("qemu-ga is not installed.")
+            self.setup_gagent_in_host(session, params, self.vm)
+            self.gagent_install(session, self.vm, *[params.get("gagent_install_cmd")])
 
-            if self._check_ga_service(session, params.get("gagent_status_cmd")):
-                logging.info("qemu-ga service is already running.")
-            else:
-                logging.info("qemu-ga service is not running.")
-                self.gagent_start(session, self.vm)
+        if self._check_ga_service(session, params.get("gagent_status_cmd")):
+            logging.info("qemu-ga service is already running.")
+        else:
+            logging.info("qemu-ga service is not running.")
+            self.gagent_start(session, self.vm)
 
-            session.close()
+        session.close()
+        if params.get("gagent_serial_type"):
             args = [params.get("gagent_serial_type"), params.get("gagent_name")]
-            self.gagent_create(params, vm, *args)
+            self.gagent_create(params, self.vm, *args)
+        else:
+            self.gagent = guest_agent.QemuAgent(self.vm, params["gagent_name"],
+                                                params["hotplug_gagent_serial_type"],
+                                                params["char_path"],
+                                                get_supported_cmds=True)
```

Comment:
Same as above, why move the execution out of the vm started judgement? If vm not started, you can't get the session. Or if you think the judgement is useless, simply remove the judgement line is fine.

Reply:
Same as above.  The if judgement should only determin whether to start vm. Otherwise if sefl.vm exists, it will not get session in previous code.